src/backend/CommandBuffer.cpp

```diff
@@ -171,6 +302,12 @@ namespace backend {
         return std::move(mIterator);
     }
 
+    std::vector<PassResourceUsage> CommandBufferBuilder::AcquirePassResourceUsage() {
+        ASSERT(!mWerePassUsagesAcquired);
+        mWerePassUsagesAcquired = true;
+        return std::move(mPassResourceUsages);
```

Comment:
uNit: "return std::move()" used here but not above in AcquireResourceUsage() where you return by value. Better to be consistent and use it in both places or neither. (I *think* it's superfluous, but I'm still getting up to speed.)

Reply:
Without the `std::move` here, a whole copy of the `PassResourceUsage` is created. `AcquireResourceUsage` doesn't need it because it takes advantage of guaranteed return value optimization.

Reply:
Out of curiosity, why is RVO guaranteed there, but not here?

Reply:
There we have the return value allocated inside the function and a single return at the end. In the calling convention, non-POD types are returned via a hidden reference: `Type f(int a)` gets converted to `void f(Type* result, int a)`. Then the compiler is able to allocate the return value directly into the "return" pointer, skipping the copy altogether. Using std::move in that case is actually a deoptimization because it forces the move constructor to be called, when it isn't needed.

For the case with std::move, the variable already exists as a member of the CommandBufferBuilder so we must have either a copy construction or a move construction (or a construction and manual copying / moving). RVO is not possible since the variable already exist, hence the move.

Reply:
Ah of course, thanks!